samples/jupyter-notebooks/ALS_python_demo.ipynb

```diff
@@ -0,0 +1,639 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "#-----------------------------------------------------------------\n",
+    "# Create kernel in SystemML's DSL using the R-like syntax for ALS\n",
+    "# Algorithms available at : https://systemml.apache.org/algorithms\n",
+    "# Below algorithm based on ALS-CG.dml\n",
+    "#-----------------------------------------------------------------\n",
+    "als_dml = \\\n",
+    "\"\"\"\n",
+    "# Default values of some parameters\n",
+    "r          = rank\n",
+    "max_iter   = 50\n",
+    "check      = TRUE\n",
+    "thr        = 0.0001\n",
+    "\n",
+    "R = table(X[,1], X[,2], X[,3])\n",
+    "\n",
+    "# check the input matrix R, if some rows or columns contain only zeros remove them from R\n",
+    "R_nonzero_ind = R != 0;\n",
+    "row_nonzeros = rowSums (R_nonzero_ind);\n",
+    "col_nonzeros = t (colSums (R_nonzero_ind));\n",
+    "orig_nonzero_rows_ind = row_nonzeros != 0;\n",
+    "orig_nonzero_cols_ind = col_nonzeros != 0;\n",
+    "num_zero_rows = nrow (R) - sum (orig_nonzero_rows_ind);\n",
+    "num_zero_cols = ncol (R) - sum (orig_nonzero_cols_ind);\n",
+    "if (num_zero_rows > 0) {\n",
+    "  print (\"Matrix R contains empty rows! These rows will be removed.\");\n",
+    "  R = removeEmpty (target = R, margin = \"rows\");\n",
+    "}\n",
+    "if (num_zero_cols > 0) {\n",
+    "  print (\"Matrix R contains empty columns! These columns will be removed.\");\n",
+    "  R = removeEmpty (target = R, margin = \"cols\");\n",
+    "}\n",
+    "if (num_zero_rows > 0 | num_zero_cols > 0) {\n",
+    "  print (\"Recomputing nonzero rows and columns!\");\n",
+    "  R_nonzero_ind = R != 0;\n",
+    "  row_nonzeros = rowSums (R_nonzero_ind);\n",
+    "  col_nonzeros = t (colSums (R_nonzero_ind));\n",
+    "}\n",
+    "\n",
+    "###### MAIN PART ######\n",
+    "\n",
+    "m = nrow (R);\n",
+    "n = ncol (R);\n",
+    "\n",
+    "# initializing factor matrices\n",
+    "U = rand (rows = m, cols = r, min = -0.5, max = 0.5);\n",
+    "M = rand (rows = n, cols = r, min = -0.5, max = 0.5);\n",
+    "\n",
+    "# initializing transformed matrices\n",
+    "Rt = t(R);\n",
+    "\n",
+    "\n",
+    "loss_init = 0.0; # only used if check is TRUE\n",
+    "if (check) {\n",
+    "  loss_init = sum (R_nonzero_ind * (R - (U %*% t(M)))^2) + lambda * (sum ((U^2) * row_nonzeros) + sum ((M^2) * col_nonzeros));\n",
+    "  print (\"----- Initial train loss: \" + loss_init + \" -----\");\n",
+    "}\n",
+    "\n",
+    "lambda_I = diag (matrix (lambda, rows = r, cols = 1));\n",
+    "it = 0;\n",
+    "converged = FALSE;\n",
+    "while ((it < max_iter) & (!converged)) {\n",
+    "  it = it + 1;\n",
+    "  # keep M fixed and update U\n",
+    "  parfor (i in 1:m) {\n",
+    "    M_nonzero_ind = t(R[i,] != 0);\n",
+    "    M_nonzero = removeEmpty (target=M * M_nonzero_ind, margin=\"rows\");\n",
```

Comment:
A comment here might help to explain that this is temporarily removing associated feature rows for any missing ratings.